### PR TITLE
Timeout Decorator for search queries

### PR DESCRIPTION
**Pull request by edgarcosta**
Hi,

This is related to #2422.

With the goal of being able to test if searches are performed in a reasonable amount of time and give a better feedback to the user whenever the query takes too long, I wrote a small script that decorates our db queries with `.max_time_ms(timeout)` and should replace the following two lines
```
nres = res.count()
res = res.skip(start).limit(count)
```
present in almost every search handling code.
I have only replaced these two lines for numberfieds, but should be easy to do the same for rest of the search fields.

try it out with: 
http://localhost:37777/NumberField/?start=0&ur_primes=2%2C3%2C5%2C7&ram_quantifier=contained&count=20

In this case, the query times out while trying to do the count:
```
2018-04-22T17:27:03.625+0000 I COMMAND  [conn4343] command numberfields.fields command: count { count: "fields", query: { ramps: { $nin: [ "2", "3", "5", "7" ] } }, maxTimeMS: 60000 } planSummary: IXSCAN { ramps: 1, degree: 1 } keyUpdates:0 writeConflicts:0 exception: operation exceeded time limit code:50 numYields:2373 reslen:89 locks:{ Global: { acquireCount: { r: 4748 } }, Database: { acquireCount: { r: 2374 } }, Collection: { acquireCount: { r: 2374 } } } protocol:op_query 60662ms
```

This isn't perfect.  But I think it is much more helpful to see:
```
Error: The search query took longer than expected! Please help us improve by reporting this error here.
```
than Google's `Error: Server Error`.
Also, if the server is overloaded, the timeout will not help us, as it only limits the CPU cumulative time.

Let me know what you think.

### CODE REVIEW

**Comment by edgarcosta** [comment]
@davidlowryduda , let me rephrase your questions:

- What is the added value of raising the ValueError?
There is no added value, and I feel I should pass the method `search_input_error` and its args, but then someone somewhere must initialize `info['err']`.

- Why does `info['err']` must be initialized?
No clue, I'm just imitating what the `SearchParser` does.
If I don't do it, the template freaks out.

**Comment by edgarcosta** [comment]
I cleaned the code a bit, and have not incorporated it into elliptic curves.
try: 
http://localhost:37777/EllipticCurve/Q/?start=0&include_cm=exclude&surj_primes=2%2C3%2C5%2C7&surj_quantifier=include&count=100

you might need to lower the timeout to for example 1s, to observe it working.

**Comment by edgarcosta** [comment]
(and I broke elliptic functions, I will fix this later on)

**Comment by edgarcosta** [comment]
The sad thing is that the wrapper cannot really handle the timeout exception, and must pass an exception outside, as the outside function `whatever_search` must still return something.

I will try to do the other pages over the next couple of days.

**Comment by edgarcosta** [comment]
@JohnCremona, @jwj61 and @AndrewVSutherland, do you agree with this idea of limiting the time to do a search  query on the website? I would like to hear from one of you before trying to integrate this function with the remainder of the LMFDB.

**Comment by JohnCremona** [comment]
Yes I do like this.  I am no expert on decorators so thought that I could not comment sensibly on the suggested code changes.  Is it not the case that gunicorn (or some aspect of our configuration) already has a timeout for requests?  So is the benefit of this change that we can catch the timeout in our own code and hence display something useful, rather than having the system timeout and displaying some unhelpful error message?

**Comment by AndrewVSutherland** [comment]
@edgarcosta Yes, I think this is a good idea.  I might suggest merging and pushing what you have to beta and prod so we can try this out in production before you make changes everywhere else (do that in a separate PR). 

I presume you are still planning to keep the gunicorn timeouts in place, but possibly make them longer?

**Comment by JohnCremona** [comment]
Are you planning any changes in gunicorn config on the warwick servers, beta and lmfdb.warwick.ac.uk?

**Comment by edgarcosta** [comment]
Very good question! 

I had set the search query timeout to 
- 60 seconds for beta (and development)
- 27 seconds for production (the reason behing this number is the 30s timeout of google's load balancer)

But now looking at your config files, it looks likes I should've just set 27 seconds for both of them.
As the gunicorn timeout for http://beta.lmfdb.org/ is 30s. 
Do you think 27 seconds is enough for development?


Regarding http://lmfdb.warwick.ac.uk, I see that the gunicorn timeout is 20s, this perhaps should be raised to 30s.

**Comment by AndrewVSutherland** [comment]
I'm happy to merge this and #2426 and push to beta where we can see it in action.  Is there any reason to hold off?

**Comment by edgarcosta** [comment]
I have just pushed some small changes.
Now the timeout is the same for both versions and it got reduced 25
seconds.
This is mostly because gunicorn's for beta is set at 30s.

Feel free to merge it.

**Comment by edgarcosta** [comment]
tests have failed, I will look into those

**Comment by edgarcosta** [comment]
The current issue is: 
- the link http://hobbes.la.asu.edu/lmfdb-14/ is timing out

On the first ran, these searches also timed out (most likely because we were running several tests at the same time):
  - http://www.lmfdb.org/EllipticCurve/Q/?rank=0&include_cm=include&torsion=2&sha=4&surj_quantifier=include&count=100
  - http://www.lmfdb.org/EllipticCurve/Q/?start=0&include_cm=include&surj_primes=2&surj_quantifier=include&optimal=on&count=100

These weren't an issue when I reran the tests. To avoid the same issue in the future, I have fixed the first one with a new index, but the second one can't be fixed with any index.
